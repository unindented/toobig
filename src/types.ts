import { Chalk } from "chalk";

export interface CommonConfig {
  reporters?: ReporterConfig[];
}

export interface ScanConfig extends CommonConfig {
  budgets: BudgetsConfig;
  cwd?: string;
}

export interface LoadConfig extends CommonConfig {
  input: string | Result[];
}

export interface BudgetsConfig {
  [path: string]: string;
}

export type ReporterConfig = string | [string, object];

export interface Result {
  readonly path: string;
  readonly size: number;
  readonly maxSize: number;
}

export interface ReturnValue {
  results: Result[];
  anyOverBudget: boolean;
}

export interface ReporterConstructor {
  new (options?: object): Reporter;
}

export interface Reporter {
  onRunStart(): Promise<void> | void;
  onResult(result: Result): Promise<void> | void;
  onRunComplete(results: Result[]): Promise<void> | void;
}

export interface OutputContext {
  colors: Chalk;
  colorOverride?: Chalk;
}

export type InputStream = NodeJS.ReadableStream;

export type OutputStream = NodeJS.WritableStream;
