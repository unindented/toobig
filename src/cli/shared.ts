import { cosmiconfigSync } from "cosmiconfig";
import { existsSync, lstatSync } from "fs-extra";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const pkg = require("../../package.json") as { name: string };

export const parserConfiguration = {
  "boolean-negation": true,
  "camel-case-expansion": false,
  "combine-arrays": false,
  "dot-notation": false,
  "duplicate-arguments-array": true,
  "flatten-duplicate-arrays": false,
  "halt-at-non-option": false,
  "negation-prefix": "no-",
  "parse-numbers": false,
  "populate--": false,
  "set-placeholder-key": false,
  "short-option-groups": false,
  "strip-aliased": false,
  "strip-dashed": false,
  "unknown-options-as-args": false,
};

export const configParser = (configPath?: string): object => {
  const explorer = cosmiconfigSync(pkg.name);
  const result = isFile(configPath)
    ? explorer.load(configPath)
    : explorer.search(configPath);

  return result != null && typeof result.config === "object"
    ? (result.config as object)
    : {};
};

const isFile = (path?: string): path is string =>
  !!(path && existsSync(path) && lstatSync(path).isFile());
