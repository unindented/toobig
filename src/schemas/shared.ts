import Joi from "@hapi/joi";

export const reportersConfigSchema = Joi.array().items(
  Joi.alternatives().try(
    Joi.string(),
    Joi.array().items(Joi.string().required(), Joi.object().required())
  )
);

export const resultsConfigSchema = Joi.array().items(
  Joi.object({
    path: Joi.string().required(),
    size: Joi.number().required(),
    maxSize: Joi.number().required(),
  }).required()
);
