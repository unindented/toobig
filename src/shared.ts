import { parse } from "url";

import axios from "axios";
import { Instance as ChalkInstance, supportsColor } from "chalk";
import { createReadStream, createWriteStream, ensureFileSync } from "fs-extra";

import { InputStream, OutputContext, OutputStream, Result } from "./types";

export { supportsColor };

export const isOverBudget = ({ size, maxSize }: Result): boolean =>
  size >= maxSize;

export const getTotalSize = (results: Result[]): number =>
  results.reduce((acc, { size }) => acc + size, 0);

export const getOutputContext = (supportsColor: boolean): OutputContext => {
  return {
    colors: new ChalkInstance(!supportsColor ? { level: 0 } : undefined),
  };
};

export const getOutputStream = (
  output: string | NodeJS.WritableStream
): OutputStream => {
  if (output === "stdout" || output === "stderr") {
    return process[output];
  }

  if (typeof output === "string") {
    ensureFileSync(output);
    return createWriteStream(output);
  }

  return output;
};

export const endOutputStream = (stream: OutputStream): void => {
  if (stream === process.stdout || stream === process.stderr) {
    return;
  }

  stream.end();
};

export const getInputStream = async (
  input: string | NodeJS.ReadableStream
): Promise<InputStream> => {
  if (input === "stdin") {
    return process.stdin;
  }

  if (typeof input === "string") {
    const { protocol } = parse(input);
    return protocol == null
      ? createReadStream(input)
      : await createAxiosStream(input);
  }

  return input;
};

const createAxiosStream = async (
  input: string
): Promise<NodeJS.ReadableStream> => {
  const { data: stream } = await axios.get<NodeJS.ReadableStream>(input, {
    responseType: "stream",
    timeout: 10000,
  });
  return stream;
};
