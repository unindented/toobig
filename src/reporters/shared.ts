import { basename, dirname, relative, resolve, sep } from "path";

import bytes from "bytes";
import terminalLink from "terminal-link";

import { getTotalSize, isOverBudget, supportsColor } from "../shared";
import { OutputContext, Result } from "../types";

export { getTotalSize, isOverBudget, supportsColor };

export const formatSize = ({ size }: Result): string => bytes(size);

export const formatMaxSize = ({ maxSize }: Result): string => bytes(maxSize);

export const formatSizeVsMaxSize = (
  result: Result,
  { colors, colorOverride }: OutputContext
): string => {
  const overMaxSize = isOverBudget(result);
  const color = colorOverride
    ? colorOverride
    : overMaxSize
    ? colors.bold.red
    : colors.dim;
  const operator = overMaxSize ? ">=" : "<";

  return color(`${formatSize(result)} ${operator} ${formatMaxSize(result)}`);
};

export const formatTotalSize = (results: Result[]): string =>
  bytes(getTotalSize(results));

export const formatPath = (
  { path }: Result,
  { colors }: OutputContext
): string => {
  const relativePath = relative(process.cwd(), path);
  const dir = dirname(relativePath);
  const base = basename(relativePath);
  return `${dir !== "." ? colors.dim(dir + sep) : ""}${colors.bold(base)}`;
};

export const linkPath = (result: Result, options: OutputContext): string => {
  const formattedPath = formatPath(result, options);
  return terminalLink(
    formattedPath,
    `file://${resolve(process.cwd(), result.path)}`,
    {
      fallback: () => formattedPath,
    }
  );
};
