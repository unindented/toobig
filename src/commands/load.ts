import { loadConfigSchema } from "../schemas";
import { getInputStream, isOverBudget } from "../shared";
import { LoadConfig, Result, ReturnValue } from "../types";

import { getCompositeReporter, reportResults } from "./shared";

export const loadAndReport = async (
  config: LoadConfig
): Promise<ReturnValue> => {
  await loadConfigSchema.validateAsync(config);

  const { input, reporters = ["default"] } = config;
  const compositeReporter = getCompositeReporter(reporters);

  const results = await loadResults({ input });
  await reportResults({ results, reporter: compositeReporter });

  const anyOverBudget = results.some(isOverBudget);
  return { results, anyOverBudget };
};

const loadResults = async ({
  input,
}: {
  input: string | Result[];
}): Promise<Result[]> => {
  if (typeof input !== "string") {
    return input;
  }

  const stream = await getInputStream(input);

  return new Promise((resolve, reject) => {
    const contents: string[] = [];

    stream.on("data", (data: string | Buffer) => {
      contents.push(data.toString());
    });

    stream.on("end", () => {
      resolve(JSON.parse(contents.join("")) as Result[]);
    });

    stream.on("error", (error: Error) => {
      reject(error);
    });
  });
};
