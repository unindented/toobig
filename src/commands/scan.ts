import { resolve as resolvePath } from "path";

import bytes from "bytes";
import glob from "fast-glob";
import getFolderSize from "get-folder-size";

import { scanConfigSchema } from "../schemas";
import { isOverBudget } from "../shared";
import {
  BudgetsConfig,
  Result,
  Results,
  ReturnValue,
  ScanConfig,
  WritableResults,
} from "../types";

import { getCompositeReporter, loadResults, reportResults } from "./shared";

export const scanAndReport = async (
  config: ScanConfig
): Promise<ReturnValue> => {
  await scanConfigSchema.validateAsync(config);

  const {
    budgets,
    cwd = process.cwd(),
    baselines,
    reporters = ["default"],
  } = config;
  const compositeReporter = getCompositeReporter(reporters);

  const scannedResults = await scanResults({ budgets, cwd });
  const baselineResults = baselines
    ? await loadResults({ results: baselines })
    : undefined;
  await reportResults({
    results: scannedResults,
    baselines: baselineResults,
    reporter: compositeReporter,
  });

  const anyOverBudget = Object.values(scannedResults).some(isOverBudget);
  return { results: scannedResults, anyOverBudget };
};

const scanResults = async ({
  budgets,
  cwd,
}: {
  budgets: BudgetsConfig;
  cwd: string;
}): Promise<Results> => {
  const results: WritableResults = {};

  for (const budget of Object.keys(budgets)) {
    const maxSize = bytes(budgets[budget]);
    const paths = await glob(budget, { cwd, onlyFiles: false });

    for (const path of paths) {
      const result: Result = {
        path: path.toString(),
        size: await getSize(resolvePath(cwd, path.toString())),
        maxSize,
      };

      if (results[path] === undefined || results[path].maxSize > maxSize) {
        results[path] = result;
      }
    }
  }

  return results;
};

const getSize = (path: string): Promise<number> =>
  new Promise((resolve, reject) => {
    getFolderSize(path, (err, size) => {
      if (err !== null) {
        reject(err);
        return;
      }
      resolve(size);
    });
  });
