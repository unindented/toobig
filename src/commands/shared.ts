import { CompositeReporter } from "../reporters";
import {
  filterResults,
  getInputStream,
  isOverBaseline,
  isOverBudget,
  isUnderBaseline,
  readInputStream,
} from "../shared";
import {
  Reporter,
  ReporterConfig,
  ReporterConstructor,
  Results,
  ReturnValues,
} from "../types";

export const getCompositeReporter = (
  configReporters: readonly ReporterConfig[]
): Reporter => {
  const reporters = configReporters.map((configReporter) => {
    let reporter: Reporter;

    if (typeof configReporter === "string") {
      const ReporterClass = resolveReporter(configReporter);
      reporter = new ReporterClass();
    } else {
      const ReporterClass = resolveReporter(configReporter[0]);
      reporter = new ReporterClass(configReporter[1]);
    }

    return reporter;
  });

  return new CompositeReporter({ reporters });
};

const resolveReporter = (reporterName: string): ReporterConstructor => {
  let module: { default: ReporterConstructor };

  try {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    module = require(`../reporters/${reporterName}`);
  } catch (err) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    module = require(reporterName);
  }

  return module.default;
};

export const loadResults = async ({
  results,
}: {
  results: string | Results;
}): Promise<Results> => {
  if (typeof results !== "string") {
    return results;
  }

  const stream = await getInputStream(results);
  const contents = await readInputStream(stream);
  return JSON.parse(contents) as Results;
};

export const reportResults = async ({
  results,
  baselines,
  reporter,
}: {
  results: Results;
  baselines?: Results;
  reporter: Reporter;
}): Promise<void> => {
  await reporter.onRunStart();

  for (const result of Object.values(results)) {
    await reporter.onResult(
      result,
      baselines ? baselines[result.path] : undefined
    );
  }

  await reporter.onRunComplete(results, baselines);
};

export const getReturnValues = ({
  results,
  baselines,
}: {
  results: Results;
  baselines?: Results;
}): ReturnValues => {
  const resultsUnderBaseline = filterResults(
    results,
    isUnderBaseline(baselines)
  );
  const resultsOverBaseline = filterResults(results, isOverBaseline(baselines));
  const resultsOverBudget = filterResults(results, isOverBudget);

  const anyUnderBaseline = Object.keys(resultsUnderBaseline).length > 0;
  const anyOverBaseline = Object.keys(resultsOverBaseline).length > 0;
  const anyOverBudget = Object.keys(resultsOverBudget).length > 0;

  return {
    results,
    anyUnderBaseline,
    anyOverBaseline,
    anyOverBudget,
  };
};
