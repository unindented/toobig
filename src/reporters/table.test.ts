jest.mock("../shared", () => ({
  ...jest.requireActual("../shared"),
  getOutputStream: jest.fn().mockName("getOutputStream"),
}));

import { Writable } from "stream";

import { reportResults } from "../commands/shared";
import { getOutputStream } from "../shared";
import { Reporter, Results } from "../types";

import { TableReporter } from ".";

const noResults: Results = {};
const resultsBaselines = require("../__fixtures__/results-baselines.json") as Results;
const resultsMultipleOver = require("../__fixtures__/results-multiple-over.json") as Results;
const resultsOneOver = require("../__fixtures__/results-one-over.json") as Results;
const resultsUnder = require("../__fixtures__/results-under.json") as Results;

describe("TableReporter", () => {
  let reporter: Reporter;
  let output: string[];

  beforeEach(() => {
    output = [];
    const stream = new Writable({
      write(chunk: Buffer | string, _encoding, callback): void {
        output.push(chunk.toString());
        callback();
      },
    });
    (getOutputStream as jest.Mock).mockReturnValue(stream);
  });

  describe.each([
    ["with default options", undefined],
    ["without color", { color: false }],
    ["without verbosity", { verbose: false }],
    ["with template `markdown`", { template: "markdown" as const }],
    ["with template `ramac`", { template: "ramac" as const }],
  ])("%s", (_desc1, options) => {
    beforeEach(() => {
      reporter = new TableReporter(options);
    });

    describe.each([
      ["with all results under budget", resultsUnder, undefined],
      ["with one result over budget", resultsOneOver, undefined],
      ["with multiple results over budget", resultsMultipleOver, undefined],
      ["with no results", noResults, undefined],
      ["with baselines", resultsUnder, resultsBaselines],
    ])("%s", (_desc2, results, baselines) => {
      beforeEach(async () => {
        await reportResults({ results, baselines, reporter });
      });

      it("writes to output stream with results", () => {
        expect(output.join("")).toMatchSnapshot();
      });
    });
  });
});
