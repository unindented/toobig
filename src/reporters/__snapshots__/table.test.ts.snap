// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableReporter with default options with all results under budget writes to output stream with results 1`] = `
"------------|--------|----------|---
 [1mPath[22m       | [1mSize[22m   | [1mMax Size[22m | [1m [22m 
------------|--------|----------|---
 [2mfoo/[22m[1mbar.js[22m | 3.97KB | 4KB      | [32mâœ“[39m 
 [2mfoo/[22m[1mbaz.js[22m | 3.97KB | 4KB      | [32mâœ“[39m 
------------|--------|----------|---
            | [1m7.94KB[22m |          |   
------------|--------|----------|---

"
`;

exports[`TableReporter with default options with multiple results over budget writes to output stream with results 1`] = `
"-------------|---------|----------|---
 [1mPath[22m        | [1mSize[22m    | [1mMax Size[22m | [1m [22m 
-------------|---------|----------|---
 [2mfoo/[22m[1mbar.js[22m  | 3.97KB  | 4KB      | [32mâœ“[39m 
 [2mfoo/[22m[1mbaz.js[22m  | 4.13KB  | 4KB      | [31mâœ—[39m 
 [2mfoo/[22m[1mqux.js[22m  | 4.16KB  | 4KB      | [31mâœ—[39m 
 [2mfoo/[22m[1mquux.js[22m | 4.13KB  | 4KB      | [31mâœ—[39m 
-------------|---------|----------|---
             | [1m16.38KB[22m |          |   
-------------|---------|----------|---

"
`;

exports[`TableReporter with default options with no results writes to output stream with results 1`] = `
"------|------|----------|---
 [1mPath[22m | [1mSize[22m | [1mMax Size[22m | [1m [22m 
------|------|----------|---
      | [1m0B[22m   |          |   
------|------|----------|---

"
`;

exports[`TableReporter with default options with one result over budget writes to output stream with results 1`] = `
"------------|--------|----------|---
 [1mPath[22m       | [1mSize[22m   | [1mMax Size[22m | [1m [22m 
------------|--------|----------|---
 [2mfoo/[22m[1mbar.js[22m | 3.97KB | 4KB      | [32mâœ“[39m 
 [2mfoo/[22m[1mbaz.js[22m | 4.13KB | 4KB      | [31mâœ—[39m 
------------|--------|----------|---
            | [1m8.09KB[22m |          |   
------------|--------|----------|---

"
`;

exports[`TableReporter with template \`markdown\` with all results under budget writes to output stream with results 1`] = `
"| [1mPath[22m       | [1mSize[22m   | [1mMax Size[22m | [1m [22m |
|------------|--------|----------|---|
| [2mfoo/[22m[1mbar.js[22m | 3.97KB | 4KB      | [32mâœ“[39m |
| [2mfoo/[22m[1mbaz.js[22m | 3.97KB | 4KB      | [32mâœ“[39m |
|            | [1m7.94KB[22m |          |   |


"
`;

exports[`TableReporter with template \`markdown\` with multiple results over budget writes to output stream with results 1`] = `
"| [1mPath[22m        | [1mSize[22m    | [1mMax Size[22m | [1m [22m |
|-------------|---------|----------|---|
| [2mfoo/[22m[1mbar.js[22m  | 3.97KB  | 4KB      | [32mâœ“[39m |
| [2mfoo/[22m[1mbaz.js[22m  | 4.13KB  | 4KB      | [31mâœ—[39m |
| [2mfoo/[22m[1mqux.js[22m  | 4.16KB  | 4KB      | [31mâœ—[39m |
| [2mfoo/[22m[1mquux.js[22m | 4.13KB  | 4KB      | [31mâœ—[39m |
|             | [1m16.38KB[22m |          |   |


"
`;

exports[`TableReporter with template \`markdown\` with no results writes to output stream with results 1`] = `
"| [1mPath[22m | [1mSize[22m | [1mMax Size[22m | [1m [22m |
|------|------|----------|---|
|      | [1m0B[22m   |          |   |


"
`;

exports[`TableReporter with template \`markdown\` with one result over budget writes to output stream with results 1`] = `
"| [1mPath[22m       | [1mSize[22m   | [1mMax Size[22m | [1m [22m |
|------------|--------|----------|---|
| [2mfoo/[22m[1mbar.js[22m | 3.97KB | 4KB      | [32mâœ“[39m |
| [2mfoo/[22m[1mbaz.js[22m | 4.13KB | 4KB      | [31mâœ—[39m |
|            | [1m8.09KB[22m |          |   |


"
`;

exports[`TableReporter with template \`ramac\` with all results under budget writes to output stream with results 1`] = `
"+------------+--------+----------+---+
| [1mPath[22m       | [1mSize[22m   | [1mMax Size[22m | [1m [22m |
|------------|--------|----------|---|
| [2mfoo/[22m[1mbar.js[22m | 3.97KB | 4KB      | [32mâœ“[39m |
| [2mfoo/[22m[1mbaz.js[22m | 3.97KB | 4KB      | [32mâœ“[39m |
|------------|--------|----------|---|
|            | [1m7.94KB[22m |          |   |
+------------+--------+----------+---+

"
`;

exports[`TableReporter with template \`ramac\` with multiple results over budget writes to output stream with results 1`] = `
"+-------------+---------+----------+---+
| [1mPath[22m        | [1mSize[22m    | [1mMax Size[22m | [1m [22m |
|-------------|---------|----------|---|
| [2mfoo/[22m[1mbar.js[22m  | 3.97KB  | 4KB      | [32mâœ“[39m |
| [2mfoo/[22m[1mbaz.js[22m  | 4.13KB  | 4KB      | [31mâœ—[39m |
| [2mfoo/[22m[1mqux.js[22m  | 4.16KB  | 4KB      | [31mâœ—[39m |
| [2mfoo/[22m[1mquux.js[22m | 4.13KB  | 4KB      | [31mâœ—[39m |
|-------------|---------|----------|---|
|             | [1m16.38KB[22m |          |   |
+-------------+---------+----------+---+

"
`;

exports[`TableReporter with template \`ramac\` with no results writes to output stream with results 1`] = `
"+------+------+----------+---+
| [1mPath[22m | [1mSize[22m | [1mMax Size[22m | [1m [22m |
|------|------|----------|---|
|      | [1m0B[22m   |          |   |
+------+------+----------+---+

"
`;

exports[`TableReporter with template \`ramac\` with one result over budget writes to output stream with results 1`] = `
"+------------+--------+----------+---+
| [1mPath[22m       | [1mSize[22m   | [1mMax Size[22m | [1m [22m |
|------------|--------|----------|---|
| [2mfoo/[22m[1mbar.js[22m | 3.97KB | 4KB      | [32mâœ“[39m |
| [2mfoo/[22m[1mbaz.js[22m | 4.13KB | 4KB      | [31mâœ—[39m |
|------------|--------|----------|---|
|            | [1m8.09KB[22m |          |   |
+------------+--------+----------+---+

"
`;

exports[`TableReporter without color with all results under budget writes to output stream with results 1`] = `
"------------|--------|----------|---
 Path       | Size   | Max Size |   
------------|--------|----------|---
 foo/bar.js | 3.97KB | 4KB      | âœ“ 
 foo/baz.js | 3.97KB | 4KB      | âœ“ 
------------|--------|----------|---
            | 7.94KB |          |   
------------|--------|----------|---

"
`;

exports[`TableReporter without color with multiple results over budget writes to output stream with results 1`] = `
"-------------|---------|----------|---
 Path        | Size    | Max Size |   
-------------|---------|----------|---
 foo/bar.js  | 3.97KB  | 4KB      | âœ“ 
 foo/baz.js  | 4.13KB  | 4KB      | âœ— 
 foo/qux.js  | 4.16KB  | 4KB      | âœ— 
 foo/quux.js | 4.13KB  | 4KB      | âœ— 
-------------|---------|----------|---
             | 16.38KB |          |   
-------------|---------|----------|---

"
`;

exports[`TableReporter without color with no results writes to output stream with results 1`] = `
"------|------|----------|---
 Path | Size | Max Size |   
------|------|----------|---
      | 0B   |          |   
------|------|----------|---

"
`;

exports[`TableReporter without color with one result over budget writes to output stream with results 1`] = `
"------------|--------|----------|---
 Path       | Size   | Max Size |   
------------|--------|----------|---
 foo/bar.js | 3.97KB | 4KB      | âœ“ 
 foo/baz.js | 4.13KB | 4KB      | âœ— 
------------|--------|----------|---
            | 8.09KB |          |   
------------|--------|----------|---

"
`;
